src/v/storage/log_reader.cc

```diff
@@ -353,7 +353,7 @@ batch_timequery(const model::record_batch& b, model::timestamp t) {
     // records in the batch have different timestamps.
     model::offset result_o = b.base_offset();
     model::timestamp result_t = b.header().first_timestamp;
-    if (b.header().first_timestamp < t) {
+    if (b.header().first_timestamp < t && !b.compressed()) {
```

Comment:
i'm not seeing how compression is connected to the issue. shouldn't the batch contents be the same if compression is used or not, and in both cases, the same logic should be applied?

Reply:
is this type of query not allowed in kafka?

Reply:
It's that our code isn't able to decompress and walk through a batch: see where `verify_iterable` is called in for each record.